src/main/java/dev/mccue/resolve/maven/MavenRepository.java

```diff
@@ -11,5 +31,172 @@ public static boolean isSnapshot(String version) {
+        Classifier classifier,
+        Extension extension
+    ) {
+        return snapshotVersioning
+            .snapshotVersions()
+            .stream()
+            .filter((SnapshotVersion v) -> {
+                return (v.classifier().equals(classifier) || v.classifier().equals(new Classifier("*"))) && 
+                       (v.extension().equals(extension) || v.extension().equals(new Extension("*")));
+            })
+            .findFirst()
+            .map(v -> v.value());//I think this is right, coursier uses an extra .filter?
+    }
+
+    private static String dirModuleName(Module module, Boolean sbtAttrStub) {
+        if (sbtAttrStub) {
+            var name = module.name().value();
+            return name;
+        } else {
+            return module.name().value();
+        }
+    }
+
+    private static Project parseRawPomSax(String str) throws SaxParsingException {
+        try {
+            return Utilities.xmlParseSax(str, new PomParser()).project();
+        } catch (SaxParsingException e) { throw e;
+        }
+    }
+
+    private static String actualRoot(String root) {
+        if (root.endsWith("/")) 
+            return root.substring(0, root.length() - 1);
+        else 
+            return root;
+    }    
+
+    public static MavenRepository apply(String root) {
+        return new MavenRepository(actualRoot(root));
+    }
+    public static MavenRepository apply(String root, Optional<Authentication> authentication) {
+        return new MavenRepository(root, authentication);
+    }
+    
+    public MavenRepository(
+        String root
+    ) {
+        this.root = root;
+        this.authentication = Optional.empty();
+        this.sbtAttrStub = true; //probably don't need this sbt thing
+    }
+
+    public MavenRepository(
+        String root,
+        Optional<Authentication> authentication
+    ) {
+        this.root = root;
+        this.authentication = authentication;
+        this.sbtAttrStub = true;
+    }
+
+    public MavenRepository(
+        String root,
+        Optional<Authentication> authentication,
+        boolean sbtAttrStub
+    ) {
+        this.root = root;
+        this.authentication = authentication;
+        this.sbtAttrStub = sbtAttrStub;
+    }
+
+    public String root() { return this.root; }
+    public Optional<Authentication> authentication() { return this.authentication; }
+
+    @Override
+    public String toString() {
+        return "MavenRepository[name=" + this.root + ", authentication=" + this.authentication + "]";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return (this == o) || (
+            o instanceof MavenRepository other &&
+            Objects.equals(this.root, other.root) &&
+            Objects.equals(this.authentication, other.authentication)
+        );
+    }
+
+    @Override
+    public int hashCode() {
+        return this.root.hashCode() * this.authentication.hashCode();
+    }
```

Comment:
Objects.hash is probably the thing to use when making a standard mechanical hashcode